hp1_re: hooks into the HP1 port code in ../hp1

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 74ab008f..335fb6e8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1136,6 +1136,9 @@ target_include_directories(SurrealVideo PRIVATE SurrealVideo)
 set_target_properties(SurrealVideo PROPERTIES CXX_STANDARD 20)
 set_target_properties(SurrealVideo PROPERTIES C_STANDARD 17)
 
+# hp1_re: Harry Potter 1 port code lives outside this subtree, in ../hp1
+include(${CMAKE_CURRENT_SOURCE_DIR}/../hp1/hp1.cmake OPTIONAL)
+
 add_library(SurrealCommon STATIC ${SURREALCOMMON_SOURCES} ${THIRDPARTY_SOURCES})
 set_target_properties(SurrealCommon PROPERTIES CXX_STANDARD 20)
 target_precompile_headers(SurrealCommon PRIVATE $<$<COMPILE_LANGUAGE:CXX>:${CMAKE_CURRENT_SOURCE_DIR}/SurrealEngine/Precomp.h>)
diff --git a/SurrealEngine/Collision/TopLevel/CollisionSystem.cpp b/SurrealEngine/Collision/TopLevel/CollisionSystem.cpp
index 051ad9f9..6efd380e 100644
--- a/SurrealEngine/Collision/TopLevel/CollisionSystem.cpp
+++ b/SurrealEngine/Collision/TopLevel/CollisionSystem.cpp
@@ -1,5 +1,7 @@
 
 #include "Precomp.h"
+#include "Engine.h" // hp1_re
+#include "HP1.h" // hp1_re
 #include "CollisionSystem.h"
 #include "Packages/Engine/Actors/UActor.h"
 #include "Packages/Engine/Actors/Brush/UMover.h"
@@ -127,6 +129,10 @@ void CollisionSystem::AddToCollision(UActor* actor)
 				extents = bbox.extents() + 0.1f; // 0.1 for numerical stability
 			}
 		}
+		else if (engine->LaunchInfo.IsHarryPotter1() && HP1::IsBoxCollider(actor)) // hp1_re: CollideType CT_Box
+		{
+			extents = HP1::BoxCollisionExtents(actor) + 0.1f;
+		}
 
 		actor->Collision.Inserted = true;
 		actor->Collision.Location = location;
diff --git a/SurrealEngine/Collision/TopLevel/OverlapTest.cpp b/SurrealEngine/Collision/TopLevel/OverlapTest.cpp
index badaaaf5..d3e4714e 100644
--- a/SurrealEngine/Collision/TopLevel/OverlapTest.cpp
+++ b/SurrealEngine/Collision/TopLevel/OverlapTest.cpp
@@ -1,5 +1,7 @@
 
 #include "Precomp.h"
+#include "Engine.h" // hp1_re
+#include "HP1.h" // hp1_re
 #include "OverlapTest.h"
 #include "Collision/BottomLevel/OverlapAABBModel.h"
 #include "Packages/Engine/Actors/UActor.h"
@@ -281,6 +283,8 @@ Array<UActor*> OverlapTester::EncroachingActors(UActor* actor)
 
 bool OverlapTester::IsOverlapping(UActor* actor1, UActor* actor2)
 {
+	if (engine->LaunchInfo.IsHarryPotter1() && HP1::IsBoxCollider(actor1) && !HP1::IsBoxCollider(actor2)) // hp1_re: CollideType CT_Box
+		return CylinderActorOverlap(to_dvec3(actor2->Location()), actor2->CollisionHeight(), actor2->CollisionRadius(), actor1);
 	return CylinderActorOverlap(to_dvec3(actor1->Location()), actor1->CollisionHeight(), actor1->CollisionRadius(), actor2);
 }
 
@@ -333,6 +337,9 @@ bool OverlapTester::SphereActorOverlap(const dvec3& sphereCenter, double sphereR
 	if (actor->Brush()) // Ignore brushes for now
 		return false;
 
+	if (engine->LaunchInfo.IsHarryPotter1() && HP1::IsBoxCollider(actor)) // hp1_re: CollideType CT_Box
+		return HP1::BoxActorOverlapSphere(actor, sphereCenter, sphereRadius);
+
 	dvec3 center = to_dvec3(actor->Location());
 	double height = actor->CollisionHeight();
 	double radius = actor->CollisionRadius();
@@ -346,6 +353,9 @@ bool OverlapTester::CylinderActorOverlap(const dvec3& origin, double cylinderHei
 	if (actor->Brush()) // Ignore brushes for now
 		return false;
 
+	if (engine->LaunchInfo.IsHarryPotter1() && HP1::IsBoxCollider(actor)) // hp1_re: CollideType CT_Box
+		return HP1::BoxActorOverlapCylinder(actor, origin, cylinderHeight, cylinderRadius);
+
 	dvec3 center = to_dvec3(actor->Location());
 	double height = actor->CollisionHeight();
 	double radius = actor->CollisionRadius();
diff --git a/SurrealEngine/Collision/TopLevel/TraceTest.cpp b/SurrealEngine/Collision/TopLevel/TraceTest.cpp
index aef803f8..5b27c846 100644
--- a/SurrealEngine/Collision/TopLevel/TraceTest.cpp
+++ b/SurrealEngine/Collision/TopLevel/TraceTest.cpp
@@ -1,5 +1,7 @@
 
 #include "Precomp.h"
+#include "Engine.h" // hp1_re
+#include "HP1.h" // hp1_re
 #include "TraceTest.h"
 #include "Collision/BottomLevel/TraceAABBModel.h"
 #include "Collision/BottomLevel/TraceRayModel.h"
@@ -249,6 +251,15 @@ void TraceTester::TraceActor(UActor* actor, const dvec3& origin, double tmin, co
 		if (!traceActors || visibilityOnly)
 			return;
 
+		if (engine->LaunchInfo.IsHarryPotter1() && HP1::IsBoxCollider(actor)) // hp1_re: CollideType CT_Box
+		{
+			vec3 normal;
+			double t = HP1::BoxActorTrace(actor, origin, tmin, dirNormalized, tmax, height, radius, normal);
+			if (t < tmax)
+				hits.push_back({ (float)t, normal, actor, nullptr, nullptr });
+			return;
+		}
+
 		double t = CylinderActorTrace(origin, tmin, dirNormalized, tmax, height, radius, actor);
 		if (t < tmax)
 		{
diff --git a/SurrealEngine/Engine.cpp b/SurrealEngine/Engine.cpp
index 34ee5181..6dcfcf7d 100644
--- a/SurrealEngine/Engine.cpp
+++ b/SurrealEngine/Engine.cpp
@@ -1,6 +1,7 @@
 
 #include "Precomp.h"
 #include "Engine.h"
+#include "HP1.h" // hp1_re
 #include "Utils/File.h"
 #include "Utils/StrTools.h"
 #include "Utils/SHA1Sum.h"
@@ -229,6 +230,8 @@ void Engine::Run()
 				ExpressionValue::Variable(&CameraLocation, vecprop),
 				ExpressionValue::Variable(&CameraRotation, rotprop)
 				});
+			if (LaunchInfo.IsHarryPotter1()) // hp1_re: Hor+ field of view
+				CameraFovAngle = HP1::ViewFovAngle(CameraFovAngle, window->GetPixelWidth(), window->GetPixelHeight());
 		}
 
 		UpdateAudio();
diff --git a/SurrealEngine/Package/PackageManager.cpp b/SurrealEngine/Package/PackageManager.cpp
index cdb045bf..4c41b7b2 100644
--- a/SurrealEngine/Package/PackageManager.cpp
+++ b/SurrealEngine/Package/PackageManager.cpp
@@ -38,6 +38,7 @@
 #include "Packages/ConSys/Events/UConEventSetFlag.h"
 #include "Packages/ConSys/Events/UConEventSpeech.h"
 #include "Packages/ConSys/Events/UConEventTrade.h"
+#include "HP1.h" // hp1_re
 #include "Packages/ConSys/Events/UConEventTransferObject.h"
 #include "Packages/ConSys/Events/UConEventTrigger.h"
 #include "Packages/ConSys/History/UConHistory.h"
@@ -1114,6 +1115,9 @@ void PackageManager::RegisterFunctions()
 	{
 		NRMusicPlayer::RegisterFunctions();
 	}
+
+	if (launchInfo.IsHarryPotter1()) // hp1_re: HP1 natives (hp1/), overriding upstream stubs
+		HP1::RegisterNatives();
 }
 
 void PackageManager::RegisterNativeClasses()
diff --git a/SurrealEngine/Packages/Engine/Actors/Pawn/UPawn_Tick.cpp b/SurrealEngine/Packages/Engine/Actors/Pawn/UPawn_Tick.cpp
index de07166c..3dd2255a 100644
--- a/SurrealEngine/Packages/Engine/Actors/Pawn/UPawn_Tick.cpp
+++ b/SurrealEngine/Packages/Engine/Actors/Pawn/UPawn_Tick.cpp
@@ -7,10 +7,13 @@
 #include "Packages/Engine/Actors/Inventory/UWeapon.h"
 #include "Utils/Logger.h"
 #include "Engine.h"
+#include "HP1.h" // hp1_re
 
 void UPawn::Tick(float elapsed)
 {
 	MoveTimer() -= elapsed;
+	if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: AvgPhysicsTime for moveToward
+		HP1::PawnPhysicsTime(this, elapsed);
 
 	if (StateFrame)
 	{
@@ -236,6 +239,9 @@ bool UPawn::TickRotateTo(const vec3& target)
 
 bool UPawn::TickMoveTo(const vec3& target)
 {
+	if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: APawn::moveToward
+		return HP1::PawnMoveToward(this, target);
+
 	if (MoveTimer() < 0.0f)
 		return true;
 
diff --git a/SurrealEngine/Packages/Engine/Actors/UActor_Animation.cpp b/SurrealEngine/Packages/Engine/Actors/UActor_Animation.cpp
index aceba9af..7d5ed5e0 100644
--- a/SurrealEngine/Packages/Engine/Actors/UActor_Animation.cpp
+++ b/SurrealEngine/Packages/Engine/Actors/UActor_Animation.cpp
@@ -6,6 +6,7 @@
 #include "VM/Frame.h"
 #include "Packages/Engine/Resources/Mesh/USkeletalMesh.h"
 #include "Engine.h"
+#include "HP1.h" // hp1_re
 
 bool UActor::HasAnim(const NameString& sequence)
 {
@@ -329,6 +330,12 @@ void UActor::LoopAnim_HP(const NameString& Sequence, std::optional<float> Rate,
 
 void UActor::TickAnimation(float elapsed)
 {
+	if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: HP1 anim channels/tweening (hp1/Anim)
+	{
+		HP1::TickAnimation(this, elapsed);
+		return;
+	}
+
 	if (StateFrame && StateFrame->LatentState == LatentRunState::FinishAnim)
 	{
 		if (!IsAnimating() || AnimFrame() >= AnimLast())
diff --git a/SurrealEngine/Packages/Engine/Actors/UActor_PhysMovingBrush.cpp b/SurrealEngine/Packages/Engine/Actors/UActor_PhysMovingBrush.cpp
index 22b4b664..429c520d 100644
--- a/SurrealEngine/Packages/Engine/Actors/UActor_PhysMovingBrush.cpp
+++ b/SurrealEngine/Packages/Engine/Actors/UActor_PhysMovingBrush.cpp
@@ -9,6 +9,7 @@
 #include "Packages/Engine/Resources/Level/ULevel.h"
 #include "Packages/Engine/Resources/Level/UModel.h"
 #include "Engine.h"
+#include "HP1.h" // hp1_re
 
 void UActor::TickMovingBrush(float elapsed)
 {
@@ -17,6 +18,9 @@ void UActor::TickMovingBrush(float elapsed)
 	UMover* mover = UObject::TryCast<UMover>(this);
 	if (mover)
 	{
+		if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: HP1 Mover shadows PhysAlpha/PhysRate (hp1/HP1Mover.cpp)
+			HP1::MoverPhysicsBegin(this);
+
 		float timeLeft = elapsed;
 		while (timeLeft > 0.0f)
 		{
@@ -71,6 +75,9 @@ void UActor::TickMovingBrush(float elapsed)
 				}
 			}
 		}
+
+		if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re
+			HP1::MoverPhysicsEnd(this);
 	}
 
 	if (elapsed > 0.0f)
diff --git a/SurrealEngine/Packages/Engine/Actors/UActor_Render.cpp b/SurrealEngine/Packages/Engine/Actors/UActor_Render.cpp
index a457b9a0..e7c52100 100644
--- a/SurrealEngine/Packages/Engine/Actors/UActor_Render.cpp
+++ b/SurrealEngine/Packages/Engine/Actors/UActor_Render.cpp
@@ -6,6 +6,8 @@
 #include "Packages/Engine/Resources/Level/UModel.h"
 #include "Packages/Engine/Resources/Mesh/USkeletalMesh.h"
 #include "Packages/Engine/Resources/Textures/UTexture.h"
+#include "Engine.h" // hp1_re
+#include "HP1.h" // hp1_re
 
 void UActor::UpdateBspInfo()
 {
@@ -19,6 +21,9 @@ void UActor::UpdateBspInfo()
 		mat4 objectToWorld = mat4::translate(Location() + PrePivot()) * Coords::Rotation(Rotation()).ToMatrix() * mat4::scale(DrawScale());
 		mat4 meshToWorld = objectToWorld * mesh->meshToObject;
 		bbox = mesh->BoundingBox.transform(meshToWorld);
+		if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: HP1 skeletal render box (hp1/Anim)
+			if (auto skel = UObject::TryCast<USkeletalMesh>(mesh))
+				bbox = HP1::GetRenderBoundingBox(this, skel);
 	}
 	else if ((dt == DT_Sprite || dt == DT_SpriteAnimOnce) && (Texture()))
 	{
diff --git a/SurrealEngine/Packages/Engine/Resources/Mesh/UAnimation.cpp b/SurrealEngine/Packages/Engine/Resources/Mesh/UAnimation.cpp
index 53982dc0..8dc63eb1 100644
--- a/SurrealEngine/Packages/Engine/Resources/Mesh/UAnimation.cpp
+++ b/SurrealEngine/Packages/Engine/Resources/Mesh/UAnimation.cpp
@@ -2,12 +2,19 @@
 #include "Precomp.h"
 #include "UAnimation.h"
 #include "Engine.h"
+#include "HP1.h" // hp1_re
 
 void UAnimation::Load(ObjectStream* stream)
 {
 	UObject::Load(stream);
 
-	if (engine->LaunchInfo.IsHarryPotter1() || engine->LaunchInfo.IsHarryPotter2())
+	if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: HP1 animation format (hp1/Anim)
+	{
+		HP1::LoadAnimation(this, stream);
+		return;
+	}
+
+	if (engine->LaunchInfo.IsHarryPotter2())
 		return; // To do: format changed. Figure out how. Maybe the arrays became skippable?
 
 	int NumRefBones = stream->ReadIndex();
diff --git a/SurrealEngine/Render/RenderSubsystem.cpp b/SurrealEngine/Render/RenderSubsystem.cpp
index f3fcef6d..19b22103 100644
--- a/SurrealEngine/Render/RenderSubsystem.cpp
+++ b/SurrealEngine/Render/RenderSubsystem.cpp
@@ -12,6 +12,7 @@
 #include "Packages/Engine/Resources/Level/UPolys.h"
 #include "VM/ScriptCall.h"
 #include "Engine.h"
+#include "HP1.h" // hp1_re
 
 RenderSubsystem::RenderSubsystem(RenderDevice* renderdevice) : Device(renderdevice)
 {
@@ -55,6 +56,9 @@ void RenderSubsystem::DrawGame(float levelTimeElapsed)
 
 	PostRender();
 
+	if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: HP1_SHOTS debug screenshots
+		HP1::OnFrameRendered(Device);
+
 	Device->Unlock(true);
 }
 
diff --git a/SurrealEngine/Render/VisibleMesh.cpp b/SurrealEngine/Render/VisibleMesh.cpp
index 7f5e8efb..75100770 100644
--- a/SurrealEngine/Render/VisibleMesh.cpp
+++ b/SurrealEngine/Render/VisibleMesh.cpp
@@ -12,6 +12,7 @@
 #include "Packages/Engine/Actors/NavigationPoint/UNavigationPoint.h"
 #include "Packages/Engine/Resources/Mesh/USkeletalMesh.h"
 #include "Packages/Engine/Resources/Level/UModel.h"
+#include "HP1.h" // hp1_re
 
 bool VisibleMesh::DrawMesh(VisibleFrame* frame, UActor* actor, bool wireframe, bool translucentPass)
 {
@@ -558,6 +559,12 @@ bool VisibleMesh::DrawLodMeshFace(VisibleFrame* frame, UActor* actor, UActor* li
 
 bool VisibleMesh::DrawSkeletalMesh(VisibleFrame* frame, UActor* actor, UActor* lightLocationActor, USkeletalMesh* mesh, const mat4& ObjectToWorld, const mat3& ObjectNormalToWorld, bool translucentPass)
 {
+	if (engine->LaunchInfo.IsHarryPotter1()) // hp1_re: HP1 skeletal pose + skinning (hp1/Anim)
+	{
+		SetupLodMeshTextures(actor, mesh);
+		return HP1::DrawSkeletalMesh(frame, actor, lightLocationActor, mesh, translucentPass);
+	}
+
 	return DrawLodMesh(frame, actor, lightLocationActor, mesh, ObjectToWorld, ObjectNormalToWorld, translucentPass);
 }
 
